### ISSUE TITLE
ne10 benchmark

### ISSUE THREAD

**Issue by lieff** [projectNe10/Ne10#165]
Hello, we made some fft benchmarks, here the results:

Lib | Speed | Comment
--- | --- | ---
ne10_fft | 10125 cycles | (asm)
libavcodec | 14375 cycles | (asm)
FFT_dit4 | 34958 cycles | 
FFT_dit4 | 26583 cycles | (without bitreverse)
tricl_fft_fft | 21833 cycles | 
CkFft | 25000 cycles | 
SuperpoweredFFT | 7375 cycles | 

Measured on Android SGS6. Just for FUI that further optimization is possible.

**Comment by joesavage** [comment]
Hey @lieff, thanks for the heads up. I can't say I'm particularly familiar with Superpowered myself, but it's interesting to hear that you're seeing better numbers from their FFT routines. Would you mind giving some further details about the benchmark you're running here?

Presumably you're running each of the routines some large number of times and then taking an average. Is the setup time (e.g.  `ne10_fft_alloc_c2c_float32_neon`) included in this? What about any time taken to transform the data? (From a cursory glance, it looks like vDSP and Superpowered use non-interleaved data formats, while Ne10 keeps the real and imaginary components interleaved.) If possible, it would be great to see the code you're using to collect these numbers.

**Comment by lieff** [comment]
Yes, test repeated many times, CPU already in hot state before test code. NE10_fft_float32.neon.s.o used directly, without whole lib. It's not me makes benchmark, so, I do not have bench project right now, but I think its relatively easy to make new one with only Superpowered and ne10.

**Comment by joesavage** [comment]
Thanks for the response. I wouldn't be surprised if at least some part of the disparity here is due to the data format itself, as Ne10's FFT routines do some extra work to keep the data in an interleaved form (e.g. using `st2` and `st4` stores), but this definitely requires some closer analysis. Thanks again for letting us know.

**Comment by lieff** [comment]
Quick debug through Superpowered shows that it uses several cases SuperpoweredFFT_NoWeights4\8, SuperpoweredFFT_OneWeightPerCall4(_128), SuperpoweredFFT_OneWeightPerIteration, SuperpoweredFFT_Final4(_128) and different function versions for aligned and not aligned input data.

**Comment by lieff** [comment]
One more observation. We use similar to ne10_radix8x4\ne10_radix4x4 intrinsic code. But our benchmarks shows that only radix4 have positive effect on performance (radix8 shows slight performance degradation) using ndk gcc 4.9 compiler. It seems too high register pressure for compiler in radix8 case, not sure how asm version behave.

**Comment by joesavage** [comment]
As a quick piece of follow-up on this, did you benchmark any FFT sizes other than 1024? We had a comment [here](https://github.com/projectNe10/Ne10/issues/150#issuecomment-279203754) indicating that Ne10 can outperform Superpowered for large input sizes on some setups, and I'm curious if larger sizes entered your consideration or are appropriate to your use case at all.

**Comment by lieff** [comment]
No, wo only use 1024, and also armv7a, aarch64 picture can be different due to registers pressure.
